src/services/meta_transaction_service.ts

```diff
@@ -0,0 +1,191 @@
+import * as _ from 'lodash';
+
+import { ASSET_SWAPPER_MARKET_ORDERS_OPTS, CHAIN_ID, LIQUIDITY_POOL_REGISTRY_ADDRESS } from '../config';
+import { ONE_GWEI, ONE_SECOND_MS, QUOTE_ORDER_EXPIRATION_BUFFER_MS, TEN_MINUTES_MS } from '../constants';
+import {
+    CalculateMetaTransactionPriceResponse,
+    CalculateMetaTransactionQuoteParams,
+    GetMetaTransactionQuoteResponse,
+} from '../types';
+import { serviceUtils } from '../utils/service_utils';
+
+export class MetaTransactionService {
+    private readonly _provider: SupportedProvider;
+    private readonly _swapQuoter: SwapQuoter;
+    private readonly _contractWrappers: ContractWrappers;
+    private readonly _web3Wrapper: Web3Wrapper;
+    private readonly _devUtils: DevUtilsContract;
+
+    constructor(orderbook: Orderbook, provider: SupportedProvider) {
+        this._provider = provider;
+        const swapQuoterOpts: Partial<SwapQuoterOpts> = {
+            chainId: CHAIN_ID,
+            expiryBufferMs: QUOTE_ORDER_EXPIRATION_BUFFER_MS,
+            liquidityProviderRegistryAddress: LIQUIDITY_POOL_REGISTRY_ADDRESS,
+        };
+        this._swapQuoter = new SwapQuoter(this._provider, orderbook, swapQuoterOpts);
+        this._contractWrappers = new ContractWrappers(this._provider, { chainId: CHAIN_ID });
+        this._web3Wrapper = new Web3Wrapper(this._provider);
+        this._devUtils = new DevUtilsContract(this._contractWrappers.contractAddresses.devUtils, this._provider);
+    }
+    public async calculateMetaTransactionPriceAsync(
+        params: CalculateMetaTransactionQuoteParams,
+    ): Promise<CalculateMetaTransactionPriceResponse> {
+        const {
+            takerAddress,
+            sellAmount,
+            buyAmount,
+            buyTokenAddress,
+            sellTokenAddress,
+            slippagePercentage,
+            excludedSources,
+        } = params;
+
+        const assetSwapperOpts = {
+            ...ASSET_SWAPPER_MARKET_ORDERS_OPTS,
+            slippagePercentage,
+            bridgeSlippage: slippagePercentage,
+            excludedSources, // TODO(dave4506): overrides the excluded sources selected by chainId
+        };
+
+        let swapQuote;
+        if (sellAmount !== undefined) {
+            swapQuote = await this._swapQuoter.getMarketSellSwapQuoteAsync(
+                buyTokenAddress,
+                sellTokenAddress,
+                sellAmount,
+                assetSwapperOpts,
+            );
+        } else if (buyAmount !== undefined) {
+            swapQuote = await this._swapQuoter.getMarketBuySwapQuoteAsync(
+                buyTokenAddress,
+                sellTokenAddress,
+                buyAmount,
+                assetSwapperOpts,
+            );
+        } else {
+            throw new Error('sellAmount or buyAmount required');
+        }
+
+        const makerAssetAmount = swapQuote.bestCaseQuoteInfo.makerAssetAmount;
+        const totalTakerAssetAmount = swapQuote.bestCaseQuoteInfo.totalTakerAssetAmount;
```

Comment:
Some of this is copy-pastad from the SwapService. Does it make sense to clean up?

Reply:
That is true, but I don't think it makes sense to "clean up" these similarities since they will diverge once we add in re-couping gas costs logic.